src/main/java/net/dv8tion/jda/internal/requests/Requester.java

```diff
@@ -139,29 +136,35 @@ public Long execute(Request<?> apiRequest)
      *         the request can be made again. This could either be for the Per-Route ratelimit or the Global ratelimit.
      *         <br>Check if globalCooldown is {@code null} to determine if it was Per-Route or Global.
      */
-    public Long execute(Request<?> apiRequest, boolean handleOnRateLimit)
+    public okhttp3.Response execute(WorkTask task, boolean handleOnRateLimit)
     {
-        return execute(apiRequest, false, handleOnRateLimit);
+        return execute(task, false, handleOnRateLimit);
     }
 
-    public Long execute(Request<?> apiRequest, boolean retried, boolean handleOnRatelimit)
+    public okhttp3.Response execute(WorkTask task, boolean retried, boolean handleOnRatelimit)
     {
-        Route.CompiledRoute route = apiRequest.getRoute();
-        Long retryAfter = rateLimiter.getRateLimit(route);
-        if (retryAfter != null && retryAfter > 0)
-        {
-            if (handleOnRatelimit)
-                apiRequest.handleResponse(new Response(retryAfter, Collections.emptySet()));
-            return retryAfter;
-        }
```

Comment:
Where did this code go? do we not need it anymore? I get that ratelimiting no longer lives in this file, so is there something that is coming before the call to `execute` in the ratelimiter?

Reply:
Im guessing yes because we pass out lastResponse instead of a ratelimit retry time

Reply:
The logic here has changed, we now simply check if `work.isDone()` in the rate-limiter to handle the 429. Essentially, whenever we complete a request, we mark the task as done and then the rate-limiter won't retry it.

Before, we used to explicitly indicate to retry using the retry-after return, which limits the capabilities and was a little confusing. Now, we simplified this to a boolean check in the rate-limiter:

```java
Response response = request.execute();
if (response != null)
    updateBucket(request.getRoute(), response);
if (!request.isDone())
    retry(request);
```